libraries/AC_Avoidance/AP_OADijkstra.cpp

```diff
@@ -0,0 +1,550 @@
+    }
+
+    // must be a fence node which start from 3rd node
+    if ((id.id_type == AP_OAVisGraph::OATYPE_FENCE_POINT) && (_short_path_data_numpoints > id.id_num + 2)) {
+        node_idx = id.id_num + 2;
+        return true;
+    }
+
+    // could not find node
+    return false;
+}
+
+// find index of node with lowest tentative distance (ignore visited nodes)
+// returns true if successful and node_idx argument is updated
+bool AP_OADijkstra::find_closest_node_idx(node_index &node_idx) const
+{
+    uint16_t lowest_idx = 0;
+    float lowest_dist = 0.0f;
+    bool lowest_found = false;
+
+    // scan through all nodes looking for closest
+    for (uint16_t i=0; i<_short_path_data_numpoints; i++) {
+        const ShortPathNode &node = _short_path_data[i];
+        if (!node.visited && (!lowest_found || node.distance_cm < lowest_dist)) {
+            lowest_found = true;
+            lowest_idx = i;
+            lowest_dist = node.distance_cm;
+        }
+    }
+
+    if (lowest_found) {
+        node_idx = lowest_idx;
+    }
+    return lowest_found;
+}
+
+// calculate shortest path from origin to destination
+// returns true on success
+// requires create_polygon_fence_with_margin and create_polygon_fence_visgraph to have been run
+// resulting path is stored in _shortest_path array as vector offsets from EKF origin
+bool AP_OADijkstra::calc_shortest_path(const Location &origin, const Location &destination)
+{
+    // convert origin and destination to offsets from EKF origin
+    Vector2f origin_NE, destination_NE;
+    if (!origin.get_vector_xy_from_origin_NE(origin_NE) || !destination.get_vector_xy_from_origin_NE(destination_NE)) {
+        return false;
+    }
+
+    // create origin and destination visgraphs of polygon points
+    update_visgraph(_source_visgraph, {AP_OAVisGraph::OATYPE_SOURCE, 0}, origin_NE, true, destination_NE);
+    update_visgraph(_destination_visgraph, {AP_OAVisGraph::OATYPE_DESTINATION, 0}, destination_NE);
+
+    // clear short_path_data array
+    _short_path_data_numpoints = 0;
+
+    // add origin and destination (node_type, id, visited, distance, distance_from_idx) to short_path_data array
+    _short_path_data[_short_path_data_numpoints] = {{AP_OAVisGraph::OATYPE_SOURCE, 0}, false, 0, 0};
+    _short_path_data_numpoints++;
+    _short_path_data[_short_path_data_numpoints] = {{AP_OAVisGraph::OATYPE_DESTINATION, 0}, false, OA_DIJKSTRA_POLYGON_VISGRAPH_INFINITY_CM, OA_DIJKSTRA_POLYGON_SHORTPATH_NOTSET_IDX};
+    _short_path_data_numpoints++;
+
+    // add fence points to short_path_data array (node_type, id, visited, distance, distance_from_idx)
+    // skip last fence point because it is the same as the first
+    for (uint8_t i=0; i<_polyfence_numpoints-1; i++) {
+        _short_path_data[_short_path_data_numpoints] = {{AP_OAVisGraph::OATYPE_FENCE_POINT, i}, false, OA_DIJKSTRA_POLYGON_VISGRAPH_INFINITY_CM, OA_DIJKSTRA_POLYGON_SHORTPATH_NOTSET_IDX};
+        _short_path_data_numpoints++;
```

Comment:
You don't need to increment in the loop, you know exactly how many points you just added.

Reply:
I'll move the increment into the line above.. from a processing time point of view I don't think it matters because we either need to add (i+_short_path_data_numpoints) or increment _short_path_data_numpoints.

Reply:
You don't need an increment. You can directly add `_polyfence_numpoints-1` outside of the for loop. Which saves you from doing it each iteration.

Reply:
well, we'd either need an addition or an increment.. so not sure there's much of a difference..

Reply:
In one case you do it for each point, so you do the increment `n` times. The other is a single add, which is faster then doing it several times.

Reply:
not quite.. what I mean is that within the for loop I'd need to have an addition like i+2 so there's an addition happening anyway.

> _short_path_data[i+2] = {{AP_OAVisGraph::OATYPE_FENCE_POINT, i}, false, OA_DIJKSTRA_POLYGON_VISGRAPH_INFINITY_CM, OA_DIJKSTRA_POLYGON_SHORTPATH_NOTSET_IDX};